### DISCUSSION TITLE
Testing `isPresented`

### DISCUSSION THREAD

**Discussion by jshier** [pointfreeco/swift-composable-architecture#3472] [General]
In addition to the well known `dismiss` dependency, there's also the less known `isPresented`, which indicates a presentation from a TCA context. Specifically, it's the value of `dismiss.dismiss`, the `DismissEffect`'s closure. However, the only public init for `DismissEffect` only takes a non-optional closure, so there's no way to test `isPresented = false`, which we use to detect when our features originate without a TCA root, a common occurrence in a mixed TCA / UIKit codebase. How can we test this logic?

**Comment by stephencelis** [comment]
Can you share how you want to go about testing this? While testing it directly with overriding `dismiss` or `isPresented` might be convenient, we wonder if testing the _integration_ would be better. _E.g._ maybe a `TestStoreOf<Child>` would exercise `isPresented == false` by default, but `TestStoreOf<Parent>` with a presented `Child` would exercise `isPresented == true`

**Reply by jshier** [reply]
This is specifically for the case where we're not in an existing TCA context, so that's not really a consideration we need. 

I'm otherwise not sure, since I'm not sure what you'd otherwise consider to be "presented" in a test context, aside from the existence of a `DismissEffect`.

**Reply by wattson12** [reply]
Adding to this thread since I was searching with the same question today: we have a codebase we are about 80% migrated to TCA / SwiftUI but have some UIKit views still. In this case we have a TCA view which can be presented from either other TCA views or UIKit. When presented from TCA we call the bundled `DismissEffect` but when called from UIKit we call our own routing code 

We have written the code to be something like 
```
if isPresented { 
    return .run { _ in await dismiss() } 
else {
    return .run { _ in await ourOwnDismissFunction() }
}
```

But since we aren't sure how to make `isPresented` false in tests we can't test this flow (as written - we will likely rewrite to make our own dismiss function optional, and have an `isPresentedFromUIKit` type helper but would be nice to avoid code which is nearly duplicated from existing helpers)
